README.md

```diff
@@ -1 +1,368 @@
+
+### 1.1.1 Containers
+
+IPLD assumes that links will be referenced against a blockstore. The resulting disconnected structures — while efficient for machines — can be cumbersome to work with, debug, inspect, and transmit in non-IPLD aware applications. [Locality of reference] also plays a role: IPLD links are no guaranteed to resolve. Data that is likely to be accessed together being kept inline avoids going out to the blockstore, which may incur a (fallible) network lookup.
+
+The ecosystem of tools for non-IPLD specific tools (JSON tools, CBOR tools, etc) is much wider than those designed for IPLD in particular. To make use of them today, a developer either needs to write custom code, or find an ad hoc method for reifying an IPLD DAG into some format without links such as JSON. Packaging data for 
+
+### 1.1.2 Codecs
+
+By referencing a specific [CID] in a link, the encoding and hash algorithm of that block are forced to be decided ahead of time. While common codecs are supported by most implementations, not all applications will support all possible encodings.
+
+The identity hash has been used for similar use cases, but this has several significant drawbacks including:
+
+- A codec and hash algorithm MUST be decided in advance
+- The resulting CID will not directly work with non-IPLD tools (at minimum there's a header)
+- Depending on the parameters other than hash, the data may be illegible without further processing
+- The identity CID is frozen by the Merkle root of the encoding structure
+
+## 1.2 Inline Links
+
+IPLD Inline Links solve for the above by adding a standard way to inline data in an IPLD document without altering its hash. Link configuration (hashing algorithm, data encoding) is inherited from the surrounding context, unless the link is present. Explicitly including the CID to provide this configuration is OPTIONAL.
+
+Inline links do not modify the IPLD data model itself, but rather extend the format in a way that can be round-tripped to and from standard IPLD as desired.
+
+# 2 Format
+
+The basic structure of an inline link is similar to how reference links are represented in the most common IPLD codecs: signalling a link by wrapping the payload in a map with the single `"/"` key. By convention, this signals an IPLD-specific extension to a base codec.
+
+## 2.1 Embed Wrapper
+
+An inline link MUST be signalled by wrapping in a map with a `"/"` key.
+
+``` ipldsch
+type EmbedWrapper struct {
+  embed InlineLink (rename "/")
+}
+```
+
+## 2.2 Embedded IPLD Payload
+
+The inlined DAG payload MUST contain the inlined DAG. The CID MAY be present or set to `Null`.
+
+``` ipldsch
+type Embed struct {
+  link nullable Link
+  data          Any
+}
+```
+
+### 2.2.1 Explicit
+
+``` js
+{
+  "name": "Alonzo Church",
+  "birthday": {
+    "/": {
+      "link": "bafyreif7dowvi5nuzzijawl22vpqsughufapj455diyflrk7htswzbjid4", // DAG-CBOR & SHA2-256
```

Comment:
Yeah, I agree. Your proposal basically cuts a feature that we could add later if we find that it's painful without it — much harder to remove it later than add it in later.

Let's have as small a surface area as possible 👍 I'll make a sub-PR to see what that looks like.

Reply:
Hmm if we have this, then why even include the hash? We can always calculate it directly from the data, and presumably we want to do that check anyway. I'm going to try a version without a hash _at all_ and let's see how that feels.

Reply:
> Hmm if we have this, then why even include the hash?

It could allow you incremental verification, but then most IPLD encodings do not support partial or streaming decode so probably it will not be very useful in practice.